[#45] 공연장 정보 조회 기능 구현

hanwix2 opened the pull request:
#45 

## 기능 개요
1. **공연장 관리 페이지에서 모든 공연장 정보 조회**
    - 관리자 권한
2. **공연장 리스트에서 특정 공연장을 선택했을 때 해당 공연장 상세 정보 확인**
    - 해당 공연장의 공연 홀 정보도 반환 (별도의 이슈에서 진행)
3. **캐싱 적용**
    - 빈번히 요청되는 데이터를 캐시로 저장하여 동일한 요청이 왔을 때 빠르게 응답
    - 적용 데이터: 공연장 정보
4. **모든 공연장 정보 조회 기능에 페이징 적용**
    - 공연장 정보가 많을 경우 모든 공연장 정보를 반환할 때 문제 발생
    - 페이지를 나눠 특정 범위 만큼만 해당 페이지의 공연장 리스트 반환

<br>

## Progress
1. **모든 공연장 정보 조회**
    - **VenueController** - GET 메소드로 DB에 저장되어 있는 모든 공연장 정보를 리턴
    - **VenueService, VenueMapper** - getAllVenues: venue 테이블의 모든 공연장 정보 리턴
    - Venue 객체를 Request 용과 Response 용 객체로 구분

2. **특정 공연장 정보 조회 기능**
    - venueId를 pathVariable로 받아 해당 공연장 정보를 반환

3. **캐싱 기능 구현**
    - 글로벌 캐싱 적용:
        - RedisCacheConfig 클래스에서 캐싱 관련 설정
        - redis를 캐시 저장소로 사용 (port: 6380 / 기존 세션 저장소는 port: 6379)
    - **Jackson**을 이용하여 Serialization 수행 (GenericJackson2JsonRedisSerializer)
    - 캐싱 적용 데이터: 공연장 정보 
        - 공연장 정보는 거의 바뀌지 않는다. 
        - durationTime: 2 hours
        - VenueResponse 클래스의 필드에 대해 final 접근제어자 삭제
            - Desrialization을 수행하기 위해선 VenueResponse 객체를 기본 생성자를 통해서 생성 필요

4. **모든 공연장 정보 조회 기능에 페이징 적용**
    - 공연장 리스트 조회 페이지 번호를 RequestParam으로 전달
    - 페이지 번호를 명시하지 않을 시 default 값 1 설정 
    - 페이지 값이 유효하지 않은 요청일 경우 1 페이지 반환

5. **Test**
    - 로컬에 따라 DB에 저장된 데이터가 상이하므로 테스트는 Postman으로 진행
    - 테스트 코드 미작성

hanwix2 committed:
[#45] 모든 공연장 정보 조회 기능 구현

- VenueController - GET 메소드로 DB에 저장되어 있는 모든 공연장 정보를 리턴
- VenueService, VenueMapper - getAllVenues: venue 테이블의 모든 공연장 정보 리턴
- Venue 객체를 Request 용과 Response 용 객체로 구분

hanwix2 committed:
[#45] 특정 공연장 정보 조회 기능과 캐싱 기능 구현

- 특정 공연장 정보 조회 기능:
    - GET - venueId를 pathVariable로 받아 해당 공연장 정보를 반환
- 글로벌 캐싱 적용:
    - redis를 캐시 저장소로 사용(port: 6380 / 기존 세션 저장소는 port: 6379)
    - Jackson을 이용하여 Serialization 수행 (GenericJackson2JsonRedisSerializer)
    - 캐싱 적용 데이터: 공연장 정보 (거의 바뀌지 않는 데이터, durationTime: 2 hours)

f-lab-dev commented on src/main/java/com/show/showticketingservice/config/RedisCacheConfig.java:
```diff
@@ -0,0 +1,48 @@
+    }
+
+    @Bean
+    public RedisCacheManager redisCacheManager(@Qualifier("redisCacheConnectionFactory") RedisConnectionFactory redisConnectionFactory) {
```
`CacheManager`는 어떤 것일까요? 주석에 한번 달아주셔도 좋을 것 같아요~

f-lab-dev commented on src/main/java/com/show/showticketingservice/config/RedisCacheConfig.java:
```diff
@@ -0,0 +1,48 @@
+                .serializeValuesWith(RedisSerializationContext.SerializationPair.fromSerializer(new GenericJackson2JsonRedisSerializer()));
+
+        Map<String, RedisCacheConfiguration> redisCacheConfigMap = new HashMap<>();
+        redisCacheConfigMap.put(CacheConstant.VENUE, redisCacheConfig.entryTtl(Duration.ofHours(2L)));
```
2시간으로 걸어주신 이유가 있으실까요?

hanwix2 replied:
공연장 정보는 웬만하면 잘 변하지 않는 정보이기 때문에 만료 시간을 짧게(분 단위) 할 필요는 없다고 생각했습니다.
그렇다고 너무 길게(24시간 이상) 잡으면 캐시 저장소를 유연하게 사용하지 못하지 않을까 싶고요.
그래서 주관적으로 판단하기에 2시간이 적절하다 싶어 이렇게 설정하였습니다.

hanwix2 replied:
추가적으로 말씀을 드리자면, 
신학님과 회의 중 공연장 정보 조회 기능은 관리자만 사용하여 빈번하게 요청이 오진 않을 것 같다고 판단하였습니다. 

그래서 현재에는 공연장 정보에 대해 캐싱을 적용하였지만 나중에는 캐싱 대상에서 제외하고 
추후 개발 될 **공연 정보**에 대해 적용 할 예정입니다.

f-lab-dev commented on src/main/resources/mybatis/mappers/VenueMapper.xml:
```diff
@@ -24,4 +24,15 @@
 
+    <select id="getAllVenues" resultType="com.show.showticketingservice.model.venue.VenueResponse">
+        SELECT id, name, address, tel, homepage
+        FROM venue
```
DB에 있는 모든 값을 불러오게 될 것 같은데 이것 전체를 불러오시는 이유가 있으실까요?

hanwix2 replied:
아직 위키에 관리자 회원에 대한 내용이 완성이 되지 않았는데요, 
관리자 회원이 **공연장 정보 관리** 페이지에서 모든 공연장에 대한 정보를 리스트로 확인합니다. 

![공연장 관리](https://user-images.githubusercontent.com/70831990/113270523-85e5f000-9314-11eb-8457-22dc691a36d9.png)

f-lab-dev replied:
공연장이 아주 많아지게되면 좀 애매해질 것 같긴한데 페이징을 넣어주면 어떨까요?

hanwix2 committed:
[#45] RedisCacheConfig에 캐싱 관련 설명 주석으로 명시

- CacheManager, RedisCacherManger, CacheResolver 설명

hanwix2 committed:
[#45] 공연장 리스트 조회 기능에 페이징 적용

- 개요: 모든 공연장 정보를 반환하지 않고 페이징을 적용하여 특정 범위 만큼만 공연장 리스트 반환
- VenueController (getVenueList 메소드):
    - 공연장 리스트 조회 페이지 번호를 PathVariable로 전달
    - 페이지 번호를 명시하지 않을 시 default 값 1 설정
- VenuePagingCriteria:
    - 페이징 적용을 위한 설정 값(현재 페이지, 페이지당 venue 리스트 수, 탐색 위치) 지정
- VenueListResponse:
    - 반환되는 객체
    - 전체 페이지 수와 지정 페이지의 공연장 정보 리스트 포함
- VenueService (getVenueList 메소드):
    - 전달받은 페이지를 기반으로 유효한 요청인지 판단하여 공연장 리스트 반환
    - 유효한 요청이 아니라면 VenueListPageutOfRangeException 예외
- VenueMapper:
    - 공연장 정보 전체 수 반환, 페이징이 적용된 공연장 정보 반환 로직 추가

f-lab-dev commented on src/main/java/com/show/showticketingservice/controller/VenueController.java:
```diff
@@ -41,10 +43,10 @@ public void deleteVenue(@PathVariable int venueId) {
     }
 
-    @GetMapping
+    @GetMapping(value = {"/list", "/list/{page}"})
```
list라는 뜻은 이미 `venues`라는 URL에 포함되어 있습니다~ 보통 페이지 같은 검색조건은 RequestParam으로 넘깁니다

hanwix2 replied:
RequestParam이 아닌 PathVariable로 페이지 번호를 넘기려다 보니 URL이 겹쳐서 list 를 사용하였네요. 
말씀하신대로 수정하겠습니다!

f-lab-dev commented on src/main/java/com/show/showticketingservice/controller/VenueController.java:
```diff
@@ -41,10 +43,10 @@ public void deleteVenue(@PathVariable int venueId) {
     @UserAuthenticationNecessary(role = AccessRoles.MANAGER)
-    public List<VenueResponse> getAllVenues() {
-        return venueService.getAllVenues();
+    public VenueListResponse getVenueList(@PathVariable Optional<Integer> page) {
```
여기는 굳이 `Optional`을 쓸 필요가 없어보이는데 쓰신 이유가 있으실까요?

hanwix2 replied:
페이지 값이 명시되지 않았을 때 기본적으로 page 를 1로 설정하기 위해 적용했습니다.
페이징을 적용하며 다른 서비스의 예시를 봤는데요. 
보통 처음 리스트 조회 요청을 할 때 첫번째 페이지가 나올텐데 1이라는 값을 처음부터 URL에 값을 명시하지 않더군요.

hanwix2 replied:
고민을 해보니 `Optional`을 사용함으로 연산에 따른 성능저하가 발생할 수도 있겠네요. 
RequestParam으로 page 값을 받을 때 defaultValue 를 설정하면 간단하겠네요.

f-lab-dev commented on src/main/java/com/show/showticketingservice/service/VenueService.java:
```diff
@@ -41,12 +44,35 @@ public void deleteVenue(int venueId) {
-    public List<VenueResponse> getAllVenues() {
-        return venueMapper.getAllVenues();
+    public VenueListResponse getVenueList(int page) {
+        if (page <= 0)
```
일반적으로 페이지 파라미터를 잘못입력하면 에러로 처리하기보단 1페이지를 보여주곤 합니다~

f-lab-dev commented on src/main/java/com/show/showticketingservice/service/VenueService.java:
```diff
@@ -41,12 +44,35 @@ public void deleteVenue(int venueId) {
+                throw new VenueListPageOutOfRangeException();
+        }
+
+        return new VenueListResponse(venueTotalPage, venueResponseList);
```
만약 venueTotalCount가 0이면 `venueResponseList`가 null일 것 같은데 차라리 빈 리스트를 리턴해주면 어떨까요?

hanwix2 replied:
`venueResponseList `를 선언할 때 `null`로 할 지 `Collections.emptyList() `로 할 지 고민하다가 null로 하긴 했는데요.
혹시 null을 보내지 말아야하는 이유가 있을까요?

f-lab-dev commented on src/main/resources/mybatis/mappers/VenueMapper.xml:
```diff
@@ -24,9 +24,11 @@
+            resultType="com.show.showticketingservice.model.venue.VenueResponse">
         SELECT id, name, address, tel, homepage
         FROM venue
+        LIMIT #{startPage}, #{amount}
```
이건 슬랙에서 얘기한대로 커서 방식 페이징과 offset 방식 페이징의 트레이드오프를 고려해주세요~

hanwix2 committed:
[#45] 공연장 리스트 조회 url 수정

- 변경된 URL: '/venues'
- page 값을 RequestParam으로 전달

hanwix2 committed:
[#45] 공연장 조회 범위를 넘어가는 요청 처리 로직 수정

- 전달된 페이지 값이 유효 범위를 넘어갈 경우, 에러로 처리하지 않고 첫번째 페이지 반환
- VenueListPageOutOfRangeException 에러 삭제

tlsgkr7416 commented on src/main/java/com/show/showticketingservice/config/RedisCacheConfig.java:
```diff
@@ -31,6 +31,20 @@ public RedisConnectionFactory redisCacheConnectionFactory() {
     }
 
+    /*
+        CacheManager:
+        - 스프링의 중앙 캐시 관리자 SPI(Service Provider Interface)로, 명명된 캐시 영역을 가져온다.
```
CacheManager에 대해 좀 더 추가하자면 redis, ConcurrentHashMap 등 캐시와 관련된 기술을 동일한 사용법으로 사용할 수 있도록 추상화 시켜준 인터페이스도 추가해 주면 더 좋을 거 같네요. 즉 PSA 개념과 관련돼서요.

tlsgkr7416 replied:
아 SPI를 다시 찾아보니까 PSA와 SPI가 비슷한 내용이여서 저 내용만으로도 괜찮겠네요.

f-lab-dev commented:
빈 리스트 리턴쪽 한번 고려해주세요~

hanwix2 committed:
[#45] 공연장 조회 전체 페이지 수 기본값 수정

- 기존 0에서 1로 수정
- 기본적으로 보여지는 리스트의 페이지수는 아무 데이터가 없어도 1부터 시작

hanwix2 committed:
Merge branch 'develop' of https://github.com/f-lab-edu/show-ticketing-service into feature/45/jh
